otherlibs/unix/unix.ml

```diff
@@ -938,65 +943,47 @@ let rec waitpid_non_intr pid =
 
 external sys_exit : int -> 'a = "caml_sys_exit"
 
+external fork_and_execvp:
+  string -> string array -> file_descr -> file_descr -> file_descr -> int =
+    "unix_fork_and_execvp"
+
+external fork_and_execvpe:
+  string -> string array -> string array -> file_descr -> file_descr ->
+  file_descr -> int =
+    "unix_fork_and_execvpe_bytecode"
+    "unix_fork_and_execvpe"
+
+external fork_and_execve:
+  string -> string array -> string array -> file_descr -> file_descr ->
+  file_descr -> int =
+    "unix_fork_and_execve_bytecode"
+    "unix_fork_and_execve"
+
+let create_process = fork_and_execvp
+
+let create_process_env path args envs stdin stdout stderr =
+  try
+    fork_and_execvpe path args envs stdin stdout stderr
+  with
+  | Unix_error(ENOSYS, _, _) ->
```

Comment:
So this is where a difference in error case behavior may be observable in downstream users.

On platforms that don't have `execvpe(2)`, e.g. macOS and other BSD variants, the `ENOSYS` exception case is used, and accordingly the argument translation using `ck_execvpe` and dispatch to `fork_and_execve` has the effect that failure to locate the executable image in the path happens before the call to `fork(2)` rather than afterward and prior to the call to `execve(2)`. Specifically, the old logic would iterate in the child process looking for a successful call to `execve(2)` if the first call returned an `ENOEXEC` error. The new logic does not attempt to call `execve(2)` more than once.

Accordingly, the user visible behavior change is that `Unix.create_process_env` can exception where it would previously succeed, _in the case where searching the path finds a valid executable later in the search after an earlier non-executable one is found._

Reply:
Worth noting: I suspect the current behavior of `Unix.create_process_env` may be associated with security considerations, and I think an argument could be made that the change I am proposing could be regarded as a security improvement.